library/Graphite/Web/Widget/Graphs/Icingadb/IcingadbGraph.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+/* Icinga Graphite Web | (c) 2022 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Graphite\Web\Widget\Graphs\Icingadb;
+
+use Icinga\Module\Icingadb\Widget\EmptyState;
+use Icinga\Web\Url;
+use Icinga\Module\Icingadb\Model\Host;
+use InvalidArgumentException;
+use ipl\Html\BaseHtmlElement;
+use ipl\Html\Html;
+use ipl\Html\HtmlString;
+use ipl\Web\Widget\Link;
+
+/**
+* Base class for list items
+*/
+class IcingadbGraph extends BaseHtmlElement
+{
+    protected $defaultAttributes = ['class' => 'grid'];
+
+    /** @var Iterable */
+    protected $objects;
+
+    protected $tag = 'div';
+
+    /** @var Url */
+    protected $hostBaseUrl;
+
+    /** @var Url */
+    protected $serviceBaseUrl;
+
+    /** Create a new Graph item
+     *
+     * @param Iterable $objects
+     *
+     * @param Url $hostBaseUrl Base url for hosts
+     *
+     * @param Url|null $serviceBaseUrl Base url for services
+     *
+     */
+    public function __construct($objects, $hostBaseUrl, $serviceBaseUrl = null)
+    {
+        if (! is_iterable($objects)) {
+            throw new InvalidArgumentException('Data must be an array or an instance of Traversable');
+        }
+
+        $this->objects = $objects;
+        $this->hostBaseUrl = $hostBaseUrl;
+        $this->serviceBaseUrl = $serviceBaseUrl;
+    }
+
+    protected function assemble()
+    {
+        if ($this->objects->count() === 0) {
```

Comment:
If you'd require a `ipl\Orm\ResultSet` to be passed as `$objects`, you'd be able to call `->hasResult()` here which is more efficient. You'll get the result set by calling `->execute()` on the query. (In the controller!)